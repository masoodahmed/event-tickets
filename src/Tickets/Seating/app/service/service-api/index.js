// Get the service base URL without the trailing slash.
import { baseUrl } from './externals.js';
import {
	setIsReady,
	setEstablishingReadiness,
	registerAction,
	removeAction,
	getRegisteredActions,
	getToken,
	setToken,
	getHandlerForAction,
} from './state.js';
import {
	INBOUND_APP_READY,
	INBOUND_APP_READY_FOR_DATA,
	INBOUND_SEATS_SELECTED,
	OUTBOUND_HOST_READY,
	OUTBOUND_SEAT_TYPE_TICKETS,
} from './service-actions.js';
import { defaultMessageHandler } from './message-handlers';

/**
 * Posts a message to the service iframe.
 *
 * @since TBD
 *
 * @param {HTMLIFrameElement} iframe The iframe to post the message to.
 * @param {string}            action The message action.
 * @param {*}                 data   The message data.
 *
 * @return {void}
 */
export function sendPostMessage(iframe, action, data) {
	const token = iframe.closest('[data-token]').dataset.token;

	if (!token) {
		console.error('No token found in iframe element');
		return;
	}

	iframe.contentWindow.postMessage(
		{
			action,
			token,
			data: data || {},
		},
		baseUrl
	);
}

/**
 * The function used to handle the messages received from the service.
 *
 * All messages that not conform to the expected format, or not contain the
 * required data, will be ignored.
 * Handlers are called based on the `state.actionsMap` map that is controlled
 * using the `takeAction` and `takeEveryAction` functions.
 *
 * @since TBD
 *
 * @param {MessageEvent} event The message event received from the service.
 */
export function catchMessage(event) {
	if (
		!(
			event.origin === baseUrl &&
			event.data.token &&
			event.data.token === getToken()
		)
	) {
		return;
	}

	const action = event.data.action;

	if (!action) {
		console.error('No action found in message');
		return;
	}

	const handler = getHandlerForAction(action, defaultMessageHandler);

	handler(event.data.data);
}

/**
 * Listens for service messages.
 *
 * @since TBD
 *
 * @param {HTMLIFrameElement} iframe The iframe to listen for messages from.
 *
 * @return {void}
 */
export function startListeningForServiceMessages(iframe) {
	const tokenProvider = iframe.closest('[data-token]');

	if (!tokenProvider) {
		console.error('No token provider found in iframe element');
		return;
	}

	const token = tokenProvider.dataset.token;

	if (!token) {
		console.error('No token found in token provider element');
		return;
	}

	setToken(token);

	window.addEventListener('message', catchMessage);
}

/**
 * Starts the process of establishing the connection with the service through the iframe.
 *
 * The connection is initiated by the Service by sending a `app_postmessage_ready` message through the iframe.
 * The Site will reply with a `host_postmessage_ready` message to confirm the connection is established.
 *
 * @since TBD
 *
 * @param {HTMLIFrameElement} iframe The iframe to establish the connection with the service.
 *
 * @return {Promise<void>} A promise that will be resolved when the connection is established.
 */
export async function establishReadiness(iframe) {
	// Before setting the iframe source, start listening for messages from the service.
	startListeningForServiceMessages(iframe);

	// Build a promise that will resolve when the Service sends the ready message.
	const promise = new Promise((resolve) => {
		const acknowledge = () => {
			removeAction(INBOUND_APP_READY);

			setIsReady(true);
			setEstablishingReadiness(false);

			// Acknowledge the readiness, do not wait for a reply.
			sendPostMessage(iframe, OUTBOUND_HOST_READY);

			// Readiness is established, clear the timeout.
			clearTimeout(timeoutId);

			console.debug('Readiness established.');

			resolve();
		};

		// When the ready message from the service is received, acknowledge the readiness, resolve the promise.
		registerAction(INBOUND_APP_READY, acknowledge);
	});

	// Seat a 10s timeout to reject the promise if the connection is not established.
	const timeoutId = setTimeout(() => {
		promise.reject(new Error('Connection to service timed out'));
	}, 10000);

	// Finally start loading the service in the iframe and wait for its ready message.
	iframe.src = iframe.dataset.src;

	return promise;
}

window.tec = window.tec || {};
window.tec.tickets.seating = window.tec.tickets.seating || {};
window.tec.tickets.seating.service = {
	...(window.tec.tickets.seating.service || {}),
	INBOUND_APP_READY,
	INBOUND_APP_READY_FOR_DATA,
	OUTBOUND_HOST_READY,
	OUTBOUND_SEAT_TYPE_TICKETS,
	INBOUND_SEATS_SELECTED,
	sendPostMessage,
	startListeningForServiceMessages,
	establishReadiness,
	registerAction,
	removeAction,
	getRegisteredActions,
};
