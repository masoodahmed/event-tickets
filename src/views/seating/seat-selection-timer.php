<?php
/**
 * Seating seat selection timer template.
 *
 * Override this template in your own theme by creating a file at
 * [your-theme]/tribe/tickets-seating/seat-selection-timer.php
 *
 * @since   TBD
 *
 * @version TBD
 *
 * @var string $token        The token used to keep track of the seat selection timer.
 * @var string $redirect_url The URL to redirect the user to when the timer expires.
 * @var int    $post_id      The post ID of the post to purchase tickets for.
 */
?>

<div class="tec-tickets-seating__timer"
	 data-token="<?php echo esc_attr( $token ); ?>"
	 data-redirect-url="<?php echo esc_attr( $redirect_url ); ?>"
	 data-post-id="<?php echo esc_attr( $post_id ); ?>"
>
	<div class="dashicons dashicons-clock"></div>
	<div class="tec-tickets-seating__message">
			<span>
				<span class="tec-tickets-seating__message-text">
					<?php echo esc_html_x('Seat selections reserved for ', 'Seat selection timer text', 'event-tickets'); ?>
				</span>
				<span class="tec-tickets-seating__message-time">
					<!-- This will be set by the js component. -->
				</span>
			</span>
	</div>
</div>

