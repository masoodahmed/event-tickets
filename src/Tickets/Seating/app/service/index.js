import './errors.js';
import './notices.js';
import './service-api';
import './iframe.js';
