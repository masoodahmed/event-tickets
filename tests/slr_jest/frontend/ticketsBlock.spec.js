import {
	addModalEventListeners,
	cancelReservations,
} from '@tec/tickets/seating/frontend/ticketsBlock';
import { OUTBOUND_REMOVE_RESERVATIONS } from '@tec/tickets/seating/service';
const serviceModule = require('@tec/tickets/seating/service');

let fetchMock;

describe('Seat Selection Modal', () => {
	beforeEach(() => {
		global.fetch = jest.fn(() =>
			Promise.resolve({
				json: () => Promise.resolve({ test: 100 }),
			}),
		);
	});

	it('should listen for hide and destroy events to remove the reservations', () => {
		window['tribe-tickets-seating-modal'] = {
			on: jest.fn(),
		};

		addModalEventListeners();

		expect(window['tribe-tickets-seating-modal'].on).toHaveBeenCalledTimes(
			2
		);
		expect(window['tribe-tickets-seating-modal'].on).toHaveBeenCalledWith(
			'hide',
			cancelReservations
		);
		expect(window['tribe-tickets-seating-modal'].on).toHaveBeenCalledWith(
			'destroy',
			cancelReservations
		);
	});

	it('should not dispatch message to remove reservations if iframe not found', () => {
		serviceModule.sendPostMessage = jest.fn();
		const mockDialogElement = document.createElement('div');

		cancelReservations(mockDialogElement);

		expect(serviceModule.sendPostMessage).not.toHaveBeenCalled();
	});

	it('should dispatch message to remove reservations if iframe', () => {
		serviceModule.sendPostMessage = jest.fn();
		const mockDialogElement = document.createElement('div');
		mockDialogElement.innerHTML = `
			<div class="tec-tickets-seating__iframe-container" data-token="test-token">
				<iframe class="tec-tickets-seating__iframe"></iframe>
			</div>
		`;
		const mockIframe = mockDialogElement.querySelector(
			'.tec-tickets-seating__iframe-container iframe.tec-tickets-seating__iframe'
		);

		cancelReservations(mockDialogElement);

		expect(serviceModule.sendPostMessage).toHaveBeenCalledTimes(1);
		expect(serviceModule.sendPostMessage).toHaveBeenCalledWith(
			mockIframe,
			OUTBOUND_REMOVE_RESERVATIONS
		);
	});
});
