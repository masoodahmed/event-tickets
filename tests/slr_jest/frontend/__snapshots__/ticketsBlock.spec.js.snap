// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Seat Selection Modal selection push to backend should add tickets to selection when receiving new tickets 1`] = `
<div
  class="event-tickets"
>
  
					
  <div
    class="tec-tickets-seating__iframe-container"
    data-token="test-token"
  >
    
						
    <iframe
      class="tec-tickets-seating__iframe"
    />
    
					
  </div>
  
					
  <div
    class="tec-tickets-seating__ticket-rows"
  >
    <div
      class="tec-tickets-seating__ticket-row"
      data-price="50"
      data-seat-label="A-1"
      data-seat-type-id="uuid-seat-type-0"
      data-ticket-id="23"
    >
      
			
      <div
        class="tec-tickets-seating__seat-color"
        style="background: #00ff00"
      />
      

			
      <div
        class="tec-tickets-seating__label"
      >
        
				
        <div
          class="tec-tickets-seating__ticket-name"
        >
          Adult
        </div>
        
				
        <div
          class="tec-tickets-seating__seat-label"
        >
          A-1
        </div>
        
			
      </div>
      

			
      <div
        class="tec-tickets-seating__ticket-price"
      >
        $50
      </div>
      
		
    </div>
    <div
      class="tec-tickets-seating__ticket-row"
      data-price="50"
      data-seat-label="A-2"
      data-seat-type-id="uuid-seat-type-0"
      data-ticket-id="23"
    >
      
			
      <div
        class="tec-tickets-seating__seat-color"
        style="background: #ff0000"
      />
      

			
      <div
        class="tec-tickets-seating__label"
      >
        
				
        <div
          class="tec-tickets-seating__ticket-name"
        >
          Adult
        </div>
        
				
        <div
          class="tec-tickets-seating__seat-label"
        >
          A-2
        </div>
        
			
      </div>
      

			
      <div
        class="tec-tickets-seating__ticket-price"
      >
        $50
      </div>
      
		
    </div>
  </div>
  
					
  <div>
    
						
    <p
      class="tec-tickets-seating__total-price"
    />
    
						
    <p
      class="tec-tickets-seating__total-text"
    />
    
					
  </div>
  
					
  <div
    class="tec-tickets-seating__modal"
  >
    
						
    <button
      class="tec-tickets-seating__sidebar-control--confirm"
    >
      
							Check Out
						
    </button>
    
					
  </div>
  
				
</div>
`;
